In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/book-recommendation-dataset/Ratings.csv
/kaggle/input/book-recommendation-dataset/Users.csv
/kaggle/input/book-recommendation-dataset/classicRec.png
/kaggle/input/book-recommendation-dataset/Books.csv
/kaggle/input/book-recommendation-dataset/DeepRec.png
/kaggle/input/book-recommendation-dataset/recsys_taxonomy2.png


In [2]:
booksdf=pd.read_csv("/kaggle/input/book-recommendation-dataset/Books.csv")
usersdf=pd.read_csv("/kaggle/input/book-recommendation-dataset/Users.csv")
ratingsdf=pd.read_csv("/kaggle/input/book-recommendation-dataset/Ratings.csv")

/tmp/ipykernel_36/2744375694.py:1: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  booksdf=pd.read_csv("/kaggle/input/book-recommendation-dataset/Books.csv")


## how does the data look like

In [3]:
booksdf

ISBN                                         Book-Title  \
0       0195153448                                Classical Mythology   
1       0002005018                                       Clara Callan   
2       0060973129                               Decision in Normandy   
3       0374157065  Flu: The Story of the Great Influenza Pandemic...   
4       0393045218                             The Mummies of Urumchi   
...            ...                                                ...   
271355  0440400988                         There's a Bat in Bunk Five   
271356  0525447644                            From One to One Hundred   
271357  006008667X  Lily Dale : The True Story of the Town that Ta...   
271358  0192126040                        Republic (World's Classics)   
271359  0767409752  A Guided Tour of Rene Descartes' Meditations o...   

                 Book-Author Year-Of-Publication  \
0         Mark P. O. Morford                2002   
1       Richard Bruce Wright                2001   
2               Carlo D'Este                1991   
3           Gina Bari Kolata                1999   
4            E. J. W. Barber                1999   
...                      ...                 ...   
271355        Paula Danziger                1988   
271356            Teri Sloat                1991   
271357      Christine Wicker                2004   
271358                 Plato                1996   
271359   Christopher  Biffle                2000   

                                               Publisher  \
0                                Oxford University Press   
1                                  HarperFlamingo Canada   
2                                        HarperPerennial   
3                                   Farrar Straus Giroux   
4                             W. W. Norton &amp; Company   
...                                                  ...   
271355                   Random House Childrens Pub (Mm)   
271356                                      Dutton Books   
271357                                HarperSanFrancisco   
271358                           Oxford University Press   
271359  McGraw-Hill Humanities/Social Sciences/Languages   

                                              Image-URL-S  \
0       http://images.amazon.com/images/P/0195153448.0...   
1       http://images.amazon.com/images/P/0002005018.0...   
2       http://images.amazon.com/images/P/0060973129.0...   
3       http://images.amazon.com/images/P/0374157065.0...   
4       http://images.amazon.com/images/P/0393045218.0...   
...                                                   ...   
271355  http://images.amazon.com/images/P/0440400988.0...   
271356  http://images.amazon.com/images/P/0525447644.0...   
271357  http://images.amazon.com/images/P/006008667X.0...   
271358  http://images.amazon.com/images/P/0192126040.0...   
271359  http://images.amazon.com/images/P/0767409752.0...   

                                              Image-URL-M  \
0       http://images.amazon.com/images/P/0195153448.0...   
1       http://images.amazon.com/images/P/0002005018.0...   
2       http://images.amazon.com/images/P/0060973129.0...   
3       http://images.amazon.com/images/P/0374157065.0...   
4       http://images.amazon.com/images/P/0393045218.0...   
...                                                   ...   
271355  http://images.amazon.com/images/P/0440400988.0...   
271356  http://images.amazon.com/images/P/0525447644.0...   
271357  http://images.amazon.com/images/P/006008667X.0...   
271358  http://images.amazon.com/images/P/0192126040.0...   
271359  http://images.amazon.com/images/P/0767409752.0...   

                                              Image-URL-L  
0       http://images.amazon.com/images/P/0195153448.0...  
1       http://images.amazon.com/images/P/0002005018.0...  
2       http://images.amazon.com/images/P/0060973129.0...  
3       http://images.amazon.com/images/P/0374157065.0...  
4       http://images.ama

In [4]:
usersdf

User-ID                            Location   Age
0             1                  nyc, new york, usa   NaN
1             2           stockton, california, usa  18.0
2             3     moscow, yukon territory, russia   NaN
3             4           porto, v.n.gaia, portugal  17.0
4             5  farnborough, hants, united kingdom   NaN
...         ...                                 ...   ...
278853   278854               portland, oregon, usa   NaN
278854   278855  tacoma, washington, united kingdom  50.0
278855   278856           brampton, ontario, canada   NaN
278856   278857           knoxville, tennessee, usa   NaN
278857   278858                dublin, n/a, ireland   NaN

[278858 rows x 3 columns]

In [5]:
ratingsdf

User-ID         ISBN  Book-Rating
0         276725   034545104X            0
1         276726   0155061224            5
2         276727   0446520802            0
3         276729   052165615X            3
4         276729   0521795028            6
...          ...          ...          ...
1149775   276704   1563526298            9
1149776   276706   0679447156            0
1149777   276709   0515107662           10
1149778   276721   0590442449           10
1149779   276723  05162443314            8

[1149780 rows x 3 columns]

In [6]:
#1).How big the data is
print("dimentions of data")
print(booksdf.shape)
print(ratingsdf.shape)
print(usersdf.shape)

# #2).what is the datatype of columns
# print("datatype of data")
# print(booksdf.info())
# print(ratingsdf.info())
# print(usersdf.info())

#3).Missing values
print("Missinng values")
print(booksdf.isnull().sum())
print(ratingsdf.isnull().sum())
print(usersdf.isnull().sum())

# #4).Mathematical look of data
# print("Mathematical look")
# print(booksdf.describe)
# print(ratingsdf.describe)
# print(usersdf.describe)

#5).Dublicates
print("Dublicates")
booksdf.duplicated().sum()
ratingsdf.duplicated().sum()
usersdf.duplicated().sum()

dimentions of data
(271360, 8)
(1149780, 3)
(278858, 3)
Missinng values
ISBN                   0
Book-Title             0
Book-Author            2
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64
User-ID        0
ISBN           0
Book-Rating    0
dtype: int64
User-ID          0
Location         0
Age         110762
dtype: int64
Dublicates


0

# Popularity Based Recommender System

In [7]:
ratings_with_name=ratingsdf.merge(booksdf,on='ISBN')

In [50]:
ratings_with_name

User-ID        ISBN  Book-Rating  \
0         276725  034545104X            0   
1         276726  0155061224            5   
2         276727  0446520802            0   
3         276729  052165615X            3   
4         276729  0521795028            6   
...          ...         ...          ...   
1031131   276704  0876044011            0   
1031132   276704  1563526298            9   
1031133   276706  0679447156            0   
1031134   276709  0515107662           10   
1031135   276721  0590442449           10   

                                                Book-Title        Book-Author  \
0                                     Flesh Tones: A Novel         M. J. Rose   
1                                         Rites of Passage         Judith Rae   
2                                             The Notebook    Nicholas Sparks   
3                                           Help!: Level 1      Philip Prowse   
4        The Amsterdam Connection : Level 4 (Cambridge ...        Sue Leather   
...                                                    ...                ...   
1031131  Edgar Cayce on the Akashic Records: The Book o...  Kevin J. Todeschi   
1031132  Get Clark Smart : The Ultimate Guide for the S...       Clark Howard   
1031133  Eight Weeks to Optimum Health: A Proven Progra...        Andrew Weil   
1031134   The Sherbrooke Bride (Bride Trilogy (Paperback))  Catherine Coulter   
1031135                                  Fourth Grade Rats     Jerry Spinelli   

        Year-Of-Publication  \
0                      2002   
1                      2001   
2                      1996   
3                      1999   
4                      2001   
...                     ...   
1031131                1998   
1031132                2000   
1031133                1997   
1031134                1996   
1031135                1996   

                                                 Publisher  \
0                                         Ballantine Books   
1                                                   Heinle   
2                                             Warner Books   
3                               Cambridge University Press   
4                               Cambridge University Press   
...                                                    ...   
1031131  A.R.E. Press (Association of Research &amp; Enlig   
1031132                                   Longstreet Press   
1031133                                    Alfred A. Knopf   
1031134                                         Jove Books   
1031135                                         Scholastic   

                                               Image-URL-S  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/0155061224.0...   
2        http://images.amazon.com/images/P/0446520802.0...   
3        http://images.amazon.com/images/P/052165615X.0...   
4        http://images.amazon.com/images/P/0521795028.0...   
...                                                    ...   
1031131  http://images.amazon.com/images/P/0876044011.0...   
1031132  http://images.amazon.com/images/P/1563526298.0...   
1031133  http://images.amazon.com/images/P/0679447156.0...   
1031134  http://images.amazon.com/images/P/0515107662.0...   
1031135  http://images.amazon.com/images/P/0590442449.0...   

                                               Image-URL-M  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/0155061224.0...   
2        http://images.amazon.com/images/P/0446520802.0...   
3        http://images.amazon.com/images/P/052165615X.0...   
4        http://images.amazon.com/images/P/0521795028.0...   
...                                                    ...   
1031131  http://images.amazon.com/images/P/0876044011.0...   
1031132  http://images.amazon.com/images/P/1563526298.0...   
1031133  http://images.amazon.com/images/P/0679447156.0...   
1031134  htt

When you perform the operation ratings_with_name.groupby('Book-Title').count()['Book-Rating'], the result is a Series. In this case, num_ratings_df is a one-dimensional array-like object that holds the counts of 'Book-Rating' for each 'Book-Title'.

The rename() method for a Series does not accept the columns argument. The columns argument is specific to DataFrames, which have named columns. Since a Series does not have columns (it has an index and values), using columns will raise a TypeError.


In [13]:
num_ratings_df=ratings_with_name.groupby('Book-Title').count()['Book-Rating']


#2. Convert Series to DataFrame:
#If you want to work with a DataFrame and want to rename the resulting column, you can reset the index:
num_ratings_df = num_ratings_df.reset_index()  
num_ratings_df.columns = ['Book-Title', 'num_ratings']  # Renaming the columns as needed
# num_ratings_df.rename(columns={'Book-Ratings':'num_ratings'},inplace=True)
num_ratings_df


Book-Title  num_ratings
0        A Light in the Storm: The Civil War Diary of ...            4
1                                   Always Have Popsicles            1
2                    Apple Magic (The Collector's series)            1
3        Ask Lily (Young Women of Faith: Lily Series, ...            1
4        Beyond IBM: Leadership Marketing and Finance ...            1
...                                                   ...          ...
241066                                      Ã?Â?lpiraten.            2
241067                     Ã?Â?rger mit Produkt X. Roman.            4
241068                                Ã?Â?sterlich leben.            1
241069                              Ã?Â?stlich der Berge.            3
241070                                  Ã?Â?thique en toc            2

[241071 rows x 2 columns]

In [37]:
# Drop rows where 'Book-Ratings' is equal to the problematic string  
ratings_with_name = ratings_with_name[ratings_with_name['Book-Rating'] != '0590567330059056733005905673300590567330']  

# Optionally, if you want to ensure you're working with numeric types after dropping:  
ratings_with_name['Book-Rating'] = pd.to_numeric(ratings_with_name['Book-Rating'], errors='coerce')  


In [39]:
avg_ratings_df = ratings_with_name.groupby('Book-Title')['Book-Rating'].mean().reset_index()  
avg_ratings_df.rename(columns={'Book-Rating': 'avg_ratings'}, inplace=True)  
avg_ratings_df

Book-Title  avg_ratings
0        A Light in the Storm: The Civil War Diary of ...     2.250000
1                                   Always Have Popsicles     0.000000
2                    Apple Magic (The Collector's series)     0.000000
3        Ask Lily (Young Women of Faith: Lily Series, ...     8.000000
4        Beyond IBM: Leadership Marketing and Finance ...     0.000000
...                                                   ...          ...
241066                                      Ã?Â?lpiraten.     0.000000
241067                     Ã?Â?rger mit Produkt X. Roman.     5.250000
241068                                Ã?Â?sterlich leben.     7.000000
241069                              Ã?Â?stlich der Berge.     2.666667
241070                                  Ã?Â?thique en toc     4.000000

[241071 rows x 2 columns]

In [40]:
popular_df=num_ratings_df.merge(avg_ratings_df,on='Book-Title')
popular_df

Book-Title  num_ratings  \
0        A Light in the Storm: The Civil War Diary of ...            4   
1                                   Always Have Popsicles            1   
2                    Apple Magic (The Collector's series)            1   
3        Ask Lily (Young Women of Faith: Lily Series, ...            1   
4        Beyond IBM: Leadership Marketing and Finance ...            1   
...                                                   ...          ...   
241066                                      Ã?Â?lpiraten.            2   
241067                     Ã?Â?rger mit Produkt X. Roman.            4   
241068                                Ã?Â?sterlich leben.            1   
241069                              Ã?Â?stlich der Berge.            3   
241070                                  Ã?Â?thique en toc            2   

        avg_ratings  
0          2.250000  
1          0.000000  
2          0.000000  
3          8.000000  
4          0.000000  
...             ...  
241066     0.000000  
241067     5.250000  
241068     7.000000  
241069     2.666667  
241070     4.000000  

[241071 rows x 3 columns]

In [42]:
popular_df=popular_df[popular_df['num_ratings']>=250].sort_values('avg_ratings')
popular_df

Book-Title  num_ratings  \
234951                                        Wild Animus         2502   
94382                                        Isle of Dogs          288   
143377                                  Presumed Innocent          294   
233635                                           Whispers          286   
227701       Vinegar Hill (Oprah's Book Club (Paperback))          265   
...                                                   ...          ...   
80414    Harry Potter and the Chamber of Secrets (Book 2)          556   
80426   Harry Potter and the Order of the Phoenix (Boo...          347   
80441      Harry Potter and the Sorcerer's Stone (Book 1)          278   
80422        Harry Potter and the Goblet of Fire (Book 4)          387   
80434   Harry Potter and the Prisoner of Azkaban (Book 3)          428   

        avg_ratings  
234951     1.019584  
94382      2.000000  
143377     2.139456  
233635     2.199301  
227701     2.245283  
...             ...  
80414      5.183453  
80426      5.501441  
80441      5.737410  
80422      5.824289  
80434      5.852804  

[186 rows x 3 columns]

In [47]:
popular_df = popular_df.merge(booksdf,on='Book-Title').drop_duplicates('Book-Title')[['Book-Title','Book-Author','Image-URL-M','num_ratings','avg_ratings']]
popular_df

Book-Title        Book-Author  \
0                                          Wild Animus       Rich Shapero   
1                                         Isle of Dogs  Patricia Cornwell   
7                                    Presumed Innocent        Scott Turow   
13                                            Whispers        BELVA PLAIN   
23        Vinegar Hill (Oprah's Book Club (Paperback))   A. Manette Ansay   
..                                                 ...                ...   
724   Harry Potter and the Chamber of Secrets (Book 2)      J. K. Rowling   
727  Harry Potter and the Order of the Phoenix (Boo...      J. K. Rowling   
731     Harry Potter and the Sorcerer's Stone (Book 1)      J. K. Rowling   
735       Harry Potter and the Goblet of Fire (Book 4)      J. K. Rowling   
737  Harry Potter and the Prisoner of Azkaban (Book 3)      J. K. Rowling   

                                           Image-URL-M  num_ratings  \
0    http://images.amazon.com/images/P/0971880107.0...         2502   
1    http://images.amazon.com/images/P/0425182908.0...          288   
7    http://images.amazon.com/images/P/0446359866.0...          294   
13   http://images.amazon.com/images/P/0440216745.0...          286   
23   http://images.amazon.com/images/P/0380730138.0...          265   
..                                                 ...          ...   
724  http://images.amazon.com/images/P/0439064872.0...          556   
727  http://images.amazon.com/images/P/043935806X.0...          347   
731  http://images.amazon.com/images/P/0590353403.0...          278   
735  http://images.amazon.com/images/P/0439139597.0...          387   
737  http://images.amazon.com/images/P/0439136350.0...          428   

     avg_ratings  
0       1.019584  
1       2.000000  
7       2.139456  
13      2.199301  
23      2.245283  
..           ...  
724     5.183453  
727     5.501441  
731     5.737410  
735     5.824289  
737     5.852804  

[186 rows x 5 columns]

In [48]:
popular_df['Image-URL-M'][0]

'http://images.amazon.com/images/P/0971880107.01.MZZZZZZZ.jpg'

# Collaborative Filtering Based Recommander System

In [53]:
x=ratings_with_name.groupby('User-ID').count()['Book-Rating']>200
pdhe_likhe_users=x[x].index

In [54]:
pdhe_likhe_users

Index([   254,   2276,   2766,   2977,   3363,   4017,   4385,   6251,   6323,
         6543,
       ...
       271705, 273979, 274004, 274061, 274301, 274308, 275970, 277427, 277639,
       278418],
      dtype='int64', name='User-ID', length=811)

In [55]:
filtered_rating=ratings_with_name[ratings_with_name['User-ID'].isin(pdhe_likhe_users)]
filtered_rating

User-ID        ISBN  Book-Rating  \
1150      277427  002542730X           10   
1151      277427  0026217457            0   
1152      277427  003008685X            8   
1153      277427  0030615321            0   
1154      277427  0060002050            0   
...          ...         ...          ...   
1029357   275970  1931868123            0   
1029358   275970  3411086211           10   
1029359   275970  3829021860            0   
1029360   275970  4770019572            0   
1029361   275970  9626340762            8   

                                                Book-Title  \
1150     Politically Correct Bedtime Stories: Modern Ta...   
1151                    Vegetarian Times Complete Cookbook   
1152                                              Pioneers   
1153      Ask for May, Settle for June (A Doonesbury book)   
1154                     On a Wicked Dawn (Cynster Novels)   
...                                                    ...   
1029357  There's a Porcupine in My Outhouse: Misadventu...   
1029358                                         Die Biene.   
1029359                                     The Penis Book   
1029360                                            Musashi   
1029361  Northanger Abbey (Classic Literature with Clas...   

                           Book-Author Year-Of-Publication  \
1150                 James Finn Garner                1994   
1151                        Lucy  Moll                1995   
1152             James Fenimore Cooper                1974   
1153                     G. B. Trudeau                1982   
1154                 Stephanie Laurens                2002   
...                                ...                 ...   
1029357                   Mike Tougias                2002   
1029358  Sybil Gr�¤fin Sch�¶nfeldt                1993   
1029359                   Joseph Cohen                1999   
1029360                 Eiji Yoshikawa                1995   
1029361                    Jane Austen                1996   

                                    Publisher  \
1150                John Wiley &amp; Sons Inc   
1151                    John Wiley &amp; Sons   
1152                         Thomson Learning   
1153                      Henry Holt &amp; Co   
1154                               Avon Books   
...                                       ...   
1029357                    Capital Books (VA)   
1029358  Bibliographisches Institut, Mannheim   
1029359                              Konemann   
1029360          Kodansha International (JPN)   
1029361                 Naxos Audiobooks Ltd.   

                                               Image-URL-S  \
1150     http://images.amazon.com/images/P/002542730X.0...   
1151     http://images.amazon.com/images/P/0026217457.0...   
1152     http://images.amazon.com/images/P/003008685X.0...   
1153     http://images.amazon.com/images/P/0030615321.0...   
1154     http://images.amazon.com/images/P/0060002050.0...   
...                                                    ...   
1029357  http://images.amazon.com/images/P/1931868123.0...   
1029358  http://images.amazon.com/images/P/3411086211.0...   
1029359  http://images.amazon.com/images/P/3829021860.0...   
1029360  http://images.amazon.com/images/P/4770019572.0...   
1029361  http://images.amazon.com/images/P/9626340762.0...   

                                               Image-URL-M  \
1150     http://images.amazon.com/images/P/002542730X.0...   
1151     http://images.amazon.com/images/P/0026217457.0...   
1152     http://images.amazon.com/images/P/003008685X.0...   
1153     http://images.amazon.com/images/P/0030615321.0...   
1154     http://images.amazon.com/images/P/0060002050.0...   
...                                                    ...   
1029357  http://images.amazon.com/images/P/1931868123.0...   
1029358  http://images.amazon.com/images/P/3411086211.0...   
1029359  http://images.amazon.com/images/P/3829021860.0...   
1029360  http://images.a

In [61]:
y=filtered_rating.groupby('Book-Title').count()['Book-Rating']>=50
famous_books=y[y].index
print(famous_books)

Index(['1984', '1st to Die: A Novel', '2nd Chance', '4 Blondes',
       'A Bend in the Road', 'A Case of Need',
       'A Child Called \It\": One Child's Courage to Survive"',
       'A Civil Action', 'A Day Late and a Dollar Short', 'A Fine Balance',
       ...
       'Winter Solstice', 'Wish You Well', 'Without Remorse',
       'Wizard and Glass (The Dark Tower, Book 4)', 'Wuthering Heights',
       'Year of Wonders', 'You Belong To Me',
       'Zen and the Art of Motorcycle Maintenance: An Inquiry into Values',
       'Zoya', '\O\" Is for Outlaw"'],
      dtype='object', name='Book-Title', length=706)


In [67]:
final_ratings=filtered_rating[filtered_rating['Book-Title'].isin(famous_books)]
final_ratings.sort_values('Book-Rating',ascending=False)


User-ID        ISBN  Book-Rating  \
1150      277427  002542730X           10   
45174      11676  0525945210           10   
259560     69232  051513287X           10   
259566     69232  0515136530           10   
45059      11676  0515132020           10   
...          ...         ...          ...   
435169    115490  0804111359            0   
435170    115490  0804113475            0   
435172    115490  0805063889            0   
435188    115490  0842329110            0   
1029310   275970  1586212230            0   

                                                Book-Title  \
1150     Politically Correct Bedtime Stories: Modern Ta...   
45174    A Man Named Dave: A Story of Triumph and Forgi...   
259560        Face the Fire (Three Sisters Island Trilogy)   
259566       Key of Valor (Roberts, Nora. Key Trilogy, 3.)   
45059      Heaven and Earth (Three Sisters Island Trilogy)   
...                                                    ...   
435169                                      Secret History   
435170                                     Ladder of Years   
435172    Nickel and Dimed: On (Not) Getting By in America   
435188   Left Behind: A Novel of the Earth's Last Days ...   
1029310                                              Naked   

                Book-Author Year-Of-Publication                  Publisher  \
1150      James Finn Garner                1994  John Wiley &amp; Sons Inc   
45174       David J. Pelzer                1999               Dutton Books   
259560         Nora Roberts                2002                 Jove Books   
259566         Nora Roberts                2003                 Jove Pubns   
45059          Nora Roberts                2003                 Jove Books   
...                     ...                 ...                        ...   
435169          DONNA TARTT                1993           Ballantine Books   
435170           Anne Tyler                1997                  Ivy Books   
435172   Barbara Ehrenreich                2001         Metropolitan Books   
435188           Tim Lahaye                1995   Tyndale House Publishers   
1029310       David Sedaris                2001    Time Warner Audio Major   

                                               Image-URL-S  \
1150     http://images.amazon.com/images/P/002542730X.0...   
45174    http://images.amazon.com/images/P/0525945210.0...   
259560   http://images.amazon.com/images/P/051513287X.0...   
259566   http://images.amazon.com/images/P/0515136530.0...   
45059    http://images.amazon.com/images/P/0515132020.0...   
...                                                    ...   
435169   http://images.amazon.com/images/P/0804111359.0...   
435170   http://images.amazon.com/images/P/0804113475.0...   
435172   http://images.amazon.com/images/P/0805063889.0...   
435188   http://images.amazon.com/images/P/0842329110.0...   
1029310  http://images.amazon.com/images/P/1586212230.0...   

                                               Image-URL-M  \
1150     http://images.amazon.com/images/P/002542730X.0...   
45174    http://images.amazon.com/images/P/0525945210.0...   
259560   http://images.amazon.com/images/P/051513287X.0...   
259566   http://images.amazon.com/images/P/0515136530.0...   
45059    http://images.amazon.com/images/P/0515132020.0...   
...                                                    ...   
435169   http://images.amazon.com/images/P/0804111359.0...   
435170   http://images.amazon.com/images/P/0804113475.0...   
435172   http://images.amazon.com/images/P/0805063889.0...   
435188   http://images.amazon.com/images/P/0842329110.0...   
1029310  http://images.amazon.com/images/P/1586212230.0...   

                                               Image-URL-L  
1150     http://images.amazon.com/images/P/002542730X.0...  
45174    http://images.amazon.com/images/P/0525945210.0...  
259560   http://images.amazon.com/images/P/051513287X.0...  
259566   http://images.amazon.com/images/P/051513653

# The values in the cells are the ratings that users assigned to the books. 
For instance:
If a user has rated "1984" with a score of 9.0, that means User-ID 254 rated the book "1984" as 9.
A value of 0.0 indicates that the user has not rated that particular book.

In [70]:
pt=final_ratings.pivot_table(index='Book-Title',columns='User-ID',values='Book-Rating')
pt.fillna(0,inplace=True)
pt

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0    10.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   0.0     0.0     0.0   
1st to Die: A Novel                                    0.0     0.0     0.0   
2nd Chance                                             0.0     0.0     0.0   
4 Blondes                                              0.0     0.0     0.0   
A Bend in the Road                                     0.0     0.0     0.0   
...                                                    ...     ...     ...   
Year of Wonders                                        0.0     0.0     0.0   
You Belong To Me                                       0.0     0.0     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0     0.0     0.0   
Zoya                                                   0.0     0.0     0.0   
\O\" Is for Outlaw"                                    0.0     0.0     0.0   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   0.0  ...    10.0   
1st to Die: A Novel                                    9.0  ...     0.0   
2nd Chance                                             0.0  ...     0.0   
4 Blondes                                              0.0  ...     0.0   
A Bend in the Road                                     0.0  ...     0.0   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     0.0   
You Belong To Me                                       0.0  ...     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0   
Zoya                                                   0.0  ...     0.0   
\O\" Is for Outlaw"                                    

 # Understanding Cosine Similarity
Cosine Similarity: This metric measures the cosine of the angle between two non-zero vectors. It is often used to determine how similar two items are, regardless of their magnitude. A cosine similarity of 1 indicates that the two items are identical in terms of their feature vectors, while a cosine similarity of 0 indicates that they are orthogonal (completely dissimilar).

In [71]:
from sklearn.metrics.pairwise import cosine_similarity
similarity_scores=cosine_similarity(pt)
similarity_scores.shape

(706, 706)

In [75]:
def recommend(book_name):
    #index fetch
    index=np.where(pt.index==book_name)[0][0]
    similar_items=sorted(list(enumerate(similarity_scores[index])),key=lambda x:x[1],reverse=True)[1:5]
    
    data=[]
    for i in similar_items:
        item=[]
        temp_df=booksdf[booksdf['Book-Title']==pt.index[i[0]]]
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Book-Title'].values))
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Book-Author'].values))
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Image-URL-M'].values))
        
        data.append(item)
    return data    

In [76]:
recommend('1984')

[['Animal Farm',
  'George Orwell',
  'http://images.amazon.com/images/P/0451526341.01.MZZZZZZZ.jpg'],
 ["The Handmaid's Tale",
  'Margaret Atwood',
  'http://images.amazon.com/images/P/0449212602.01.MZZZZZZZ.jpg'],
 ['Brave New World',
  'Aldous Huxley',
  'http://images.amazon.com/images/P/0060809833.01.MZZZZZZZ.jpg'],
 ['The Vampire Lestat (Vampire Chronicles, Book II)',
  'ANNE RICE',
  'http://images.amazon.com/images/P/0345313860.01.MZZZZZZZ.jpg']]

In [77]:
pt.index[545]

"The Handmaid's Tale"

In [80]:
import pickle
pickle.dump(popular_df,open('popular.pkl','wb'))
booksdf.drop_duplicates('Book-Title')

ISBN                                         Book-Title  \
0       0195153448                                Classical Mythology   
1       0002005018                                       Clara Callan   
2       0060973129                               Decision in Normandy   
3       0374157065  Flu: The Story of the Great Influenza Pandemic...   
4       0393045218                             The Mummies of Urumchi   
...            ...                                                ...   
271354  0449906736      Flashpoints: Promise and Peril in a New World   
271356  0525447644                            From One to One Hundred   
271357  006008667X  Lily Dale : The True Story of the Town that Ta...   
271358  0192126040                        Republic (World's Classics)   
271359  0767409752  A Guided Tour of Rene Descartes' Meditations o...   

                 Book-Author Year-Of-Publication  \
0         Mark P. O. Morford                2002   
1       Richard Bruce Wright                2001   
2               Carlo D'Este                1991   
3           Gina Bari Kolata                1999   
4            E. J. W. Barber                1999   
...                      ...                 ...   
271354          Robin Wright                1993   
271356            Teri Sloat                1991   
271357      Christine Wicker                2004   
271358                 Plato                1996   
271359   Christopher  Biffle                2000   

                                               Publisher  \
0                                Oxford University Press   
1                                  HarperFlamingo Canada   
2                                        HarperPerennial   
3                                   Farrar Straus Giroux   
4                             W. W. Norton &amp; Company   
...                                                  ...   
271354                                  Ballantine Books   
271356                                      Dutton Books   
271357                                HarperSanFrancisco   
271358                           Oxford University Press   
271359  McGraw-Hill Humanities/Social Sciences/Languages   

                                              Image-URL-S  \
0       http://images.amazon.com/images/P/0195153448.0...   
1       http://images.amazon.com/images/P/0002005018.0...   
2       http://images.amazon.com/images/P/0060973129.0...   
3       http://images.amazon.com/images/P/0374157065.0...   
4       http://images.amazon.com/images/P/0393045218.0...   
...                                                   ...   
271354  http://images.amazon.com/images/P/0449906736.0...   
271356  http://images.amazon.com/images/P/0525447644.0...   
271357  http://images.amazon.com/images/P/006008667X.0...   
271358  http://images.amazon.com/images/P/0192126040.0...   
271359  http://images.amazon.com/images/P/0767409752.0...   

                                              Image-URL-M  \
0       http://images.amazon.com/images/P/0195153448.0...   
1       http://images.amazon.com/images/P/0002005018.0...   
2       http://images.amazon.com/images/P/0060973129.0...   
3       http://images.amazon.com/images/P/0374157065.0...   
4       http://images.amazon.com/images/P/0393045218.0...   
...                                                   ...   
271354  http://images.amazon.com/images/P/0449906736.0...   
271356  http://images.amazon.com/images/P/0525447644.0...   
271357  http://images.amazon.com/images/P/006008667X.0...   
271358  http://images.amazon.com/images/P/0192126040.0...   
271359  http://images.amazon.com/images/P/0767409752.0...   

                                              Image-URL-L  
0       http://images.amazon.com/images/P/0195153448.0...  
1       http://images.amazon.com/images/P/0002005018.0...  
2       http://images.amazon.com/images/P/0060973129.0...  
3       http://images.amazon.com/images/P/0374157065.0...  
4       http://images.ama

In [82]:

pickle.dump(pt,open('pt.pkl','wb'))
pickle.dump(booksdf,open('books.pkl','wb'))
pickle.dump(similarity_scores,open('similarity_scores.pkl','wb'))